File: contract-lib/src/main/scala/fssi/contract/AccountState.scala
package fssi.contract

/** the state of an account
  *
  * @param accountId id of an account
  * @param amount amount of an account
  * @param assets assets of an account
  */
case class AccountState(accountId: String, amount: BigDecimal, assets: Map[String, Array[Byte]]) {
  def withAssets(_assets: Map[String, Array[Byte]]): AccountState = copy(assets = _assets)
  def withAmount(_amount: BigDecimal): AccountState = copy(amount = _amount)

  def assetOf(key: String): Option[Array[Byte]] = assets.get(key)

  def updateAsset(key: String, data: Array[Byte]): AccountState = copy(assets = assets + (key -> data))

  lazy val bytes: Array[Byte] = {
    val assetsBytes = assets.toVector.sortBy(_._1).foldLeft(Array.emptyByteArray) {(acc, n) =>
      acc ++ n._1.getBytes("utf-8") ++ n._2
    }
    accountId.getBytes("utf-8") ++ amount.toString().getBytes("utf-8") ++ assetsBytes
  }
}

object AccountState {
  def emptyFor(accountId: String): AccountState = AccountState(accountId, BigDecimal(0), Map.empty)
}

File: contract-lib/src/main/java/fssi/contract/java/AccountState.java
package fssi.contract.java;

import fssi.contract.AccountState$;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public final class AccountState {
    private fssi.contract.AccountState inner;
    private Map<String, AssetCodec<?>> codecs;

    private AccountState() {}

    fssi.contract.AccountState toScala() {
        return inner;
    }
    static AccountState fromScala(fssi.contract.AccountState inner) {
        AccountState inst = new AccountState();
        inst.inner = inner;
        return inst;
    }

    public static AccountState newInstance(String accountId) {
        AccountState inst = new AccountState();
        inst.inner = AccountState$.MODULE$.emptyFor(accountId);
        return inst;
    }

    public <T> void registerAssetCodec(String assetName, AssetCodec<T> codec) {
        if (codecs == null) {
            codecs = new HashMap<>();
        }
        codecs.put(assetName, codec);
    }

    /**
     * add some sort of asset
     * @param assetName name
     * @param asset asset
     * @param <T> asset type
     * @return self
     * @throws NoCodecException if no codec of T was registered, it would throw exceptions
     */
    public <T> AccountState setAsset(String assetName, T asset) throws NoCodecException {
        if (codecs.containsKey(assetName)) {
            AssetCodec<T> codec = (AssetCodec<T>) codecs.get(assetName);
            inner = inner.updateAsset(assetName, codec.encode(asset));
            return this;
        }
        throw new NoCodecException("no codec found for the asset named " + assetName);
    }

    public <T> Optional<T> getAsset(String assetName) throws NoCodecException {
        if (codecs.containsKey(assetName)) {
            scala.Option<byte[]> s = inner.assetOf(assetName);
            AssetCodec<T> codec = (AssetCodec<T>) codecs.get(assetName);
            if (s.isDefined()) {
                return Optional.of(codec.decode(s.get()));
            }
            else return Optional.empty();
        }
        throw new NoCodecException("no codec found for the asset named " + assetName);
    }

}
